### PR TITLE
add hall API controller

### PR DESCRIPTION
**Pull request by LozovskiM**

### CODE REVIEW

File: Cinema system/Cinema system/Controllers/CinemasController.cs [hunk 1]
```diff
@@ -128,5 +128,149 @@ public IActionResult DeleteCinema(int cinemaId)
...
+            {
+                IEnumerable<string> messages = ModelState.Values
+                    .SelectMany(x => x.Errors)
+                    .Select(x => x.ErrorMessage);
+
+                response.IsSuccessful = false;
+                response.ErrorDescriptions.AddRange(messages);
+
+                return BadRequest(response);
+            }
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Controllers/CinemasController.cs]: I see this chunk of code repeated in several methods. You should extract this logic to a separate method.

File: Cinema system/Cinema system/Services/CinemaService.cs [hunk 2]
```diff
@@ -81,5 +83,105 @@ public void DeleteCinema(int cinemaId)
...
+                {
+                    Id = h.Id,
+                    Name = h.Name,
+                    Seats = h.Seats
+                        .Select(s => new SeatView
+                        {
+                            TypeOfSeat = s.TypeOfSeat.ToString(),
+                            Row = s.Row,
+                            Place = s.Place
+                        })
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Services/CinemaService.cs]: Do you really need all the seats for each hall when getting list of halls? What could be the use case for that?

File: Cinema system/Cinema system/Services/CinemaService.cs [hunk 3]
```diff
@@ -81,5 +83,105 @@ public void DeleteCinema(int cinemaId)
...
+                    {
+                        TypeOfSeat = (SeatType)Enum.Parse(typeof(SeatType), s.TypeOfSeat),
+                        Row = s.Row,
+                        Place = s.Place
+                    }).ToList()
+            };
+
+            var cinema = _db.Cinemas
+                 .Include(c => c.Halls)
+                 .FirstOrDefault(c => c.Id == cinemaId);
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Services/CinemaService.cs]: Why would you need cinema for adding hall? Isn't just adding hall to _db.Halls enough?

File: Cinema system/Cinema system/Services/CinemaService.cs [hunk 4]
```diff
@@ -81,5 +83,105 @@ public void DeleteCinema(int cinemaId)
...
+            _db.SaveChanges();
+
+            return newHall.Id;
+        }
+
+        public void EditHall(int hallId, HallView hall)
+        {
+            var editHall = _db.Halls
+                .Include(h => h.Seats)
+                .FirstOrDefault(h => h.Id == hallId);
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Services/CinemaService.cs]: Why first or default? Are you expecting multiple records with the same Id in database?

File: Cinema system/Cinema system/Services/CinemaService.cs [hunk 5]
```diff
@@ -81,5 +83,105 @@ public void DeleteCinema(int cinemaId)
 
             _db.SaveChanges();
         }
+
+        public bool IsHallExists(int cinemaId, int hallId)
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Services/CinemaService.cs]: Name starting with "Is..." shouldn't be used as method name. It's more suitable for object properties.
Also something wrong with the grammar here.
Better name would be "CheckHallExists"

File: Cinema system/Cinema system/Models/ViewModels/HallView.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+{
+    public class HallView
+    {
+        public int Id { get; set; }
+
+        [Required(ErrorMessage = "Name cannot be empty")]
+        [StringLength(50, ErrorMessage = "Name must be less than 50 characters")]
+        public string Name { get; set; }
+
+        [Required(ErrorMessage = "Hall cannot be empty")]
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Models/ViewModels/HallView.cs]: You should add more details on why the hall is considered empty as it's not clear you're referring to a hall with no seats.

File: Cinema system/Cinema system/Services/CinemaService.cs [hunk 7]
```diff
@@ -81,5 +83,105 @@ public void DeleteCinema(int cinemaId)
...
+            var editHall = _db.Halls
+                .Include(h => h.Seats)
+                .FirstOrDefault(h => h.Id == hallId);
+
+            editHall.Name = hall.Name;
+
+            editHall.Seats = hall.Seats
+                .Select(s => new Seat
+                {
+                    TypeOfSeat = (SeatType)System.Enum.Parse(typeof(SeatType), s.TypeOfSeat),
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Services/CinemaService.cs]: Do you need System here? That namespace already present in usings.

File: Cinema system/Cinema system/Services/CinemaService.cs [hunk 8]
```diff
@@ -81,5 +83,105 @@ public void DeleteCinema(int cinemaId)
...
+                    Row = s.Row,
+                    Place = s.Place
+                }).ToList();
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Services/CinemaService.cs]: Put .ToList() on a new line

**Review by andrei-kruhliakou-itechart — CHANGES_REQUESTED** [review]

File: Cinema system/Cinema system/Controllers/CinemasController.cs [hunk 9]
```diff
@@ -18,6 +19,13 @@ public CinemasController(ICinemaService сinemaService)
             _сinemaService = сinemaService;
         }
 
+        private IEnumerable<string> ErrorMessages(ModelStateDictionary model)
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Controllers/CinemasController.cs]: As a rule method name should start with a verb, e.g. GetErrorMessages.
Also have you considered moving this (and some other) logic to Response class?
One possible way will be defining some constructors that will take single error, model state dictionary, actual data, etc. and update IsSuccessful property accordingly.
That way you don't have to create Response object at the beginning of each action.

File: Cinema system/Cinema system/Controllers/CinemasController.cs [hunk 10]
```diff
@@ -172,18 +172,18 @@ public IActionResult GetHall(int cinemaId, int hallId)
         }
 
         [HttpPost("{cinemaId}/halls")]
-        public IActionResult CreateHall([FromBody] HallView hall, int cinemaId)
+        public IActionResult CreateHall([FromBody] HallFullView hall, int cinemaId)
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Controllers/CinemasController.cs:129]: So to create a new hall you have to fill in al the seats in it.
I'm just pointing out at this so you can think about some implications for a frontend design, no action is required here atm.

File: Cinema system/Cinema system/Services/CinemaService.cs [hunk 11]
```diff
@@ -106,35 +105,31 @@ public int CreateHall(int cinemaId, HallView hall)
...
             return newHall.Id;
         }
 
-        public void EditHall(int hallId, HallView hall)
+        public void EditHall(int hallId, HallFullView hall)
         {
-            var editHall = _db.Halls
-                .Include(h => h.Seats)
-                .FirstOrDefault(h => h.Id == hallId);
+            var editHall = _db.Halls.Find(hallId);
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Services/CinemaService.cs]: What about Include()? Is it required? Have you considered using SingleOrDefault()?

**Review by andrei-kruhliakou-itechart — CHANGES_REQUESTED** [review]

File: Cinema system/Cinema system/Controllers/CinemasController.cs [hunk 12]
```diff
@@ -14,254 +12,171 @@ public class CinemasController : ControllerBase
...
         public IActionResult GetCinemas()
         {
-            var response = new GetResponse<IEnumerable<CinemaView>>
-            {
-                IsSuccessful = true,
-                RequestedData = _сinemaService.GetCinemas()
-            };
-
-            return Ok(response);
+            return Ok(new GetResponse<IEnumerable<CinemaView>>(_сinemaService.GetCinemas()));
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Controllers/CinemasController.cs]: Here I would probably assign result of a service call to a separate variable just for better readability.
And type argument (i.e. <IEnumerable<>>) should be inferred by a compiler, so just new GetResponse() should be enough.

File: Cinema system/Cinema system/Controllers/CinemasController.cs [hunk 13]
```diff
@@ -14,254 +12,171 @@ public class CinemasController : ControllerBase
...
-                response.ErrorDescriptions.Add("This cinema does not exists.");
-                return NotFound(response);
+                return NotFound(new GetResponse<CinemaView>(ErrorOfCinemaNonexistence));
             }
-
-            response.IsSuccessful = true;
-            response.RequestedData = _сinemaService.GetCinema(cinemaId);
             
-            return Ok(response);
+            return Ok(new GetResponse<CinemaView>(_сinemaService.GetCinema(cinemaId)));
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Controllers/CinemasController.cs]: Same comment as above

File: Cinema system/Cinema system/Controllers/CinemasController.cs [hunk 14]
```diff
@@ -14,254 +12,171 @@ public class CinemasController : ControllerBase
...
-                response.IsSuccessful = false;
-                response.ErrorDescriptions.AddRange(ErrorMessages(ModelState));
-
-                return BadRequest(response);
+                return BadRequest(new CreateResponse(ModelState));
             }
 
-            response.IsSuccessful = true;
-            response.Id = _сinemaService.CreateCinema(cinema);
+            var response = new CreateResponse(_сinemaService.CreateCinema(cinema));
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Controllers/CinemasController.cs]: Same here

File: Cinema system/Cinema system/Models/Utils/Response.cs [hunk 15]
```diff
@@ -10,17 +12,50 @@ public class Response
...
+                    .Select(x => x.ErrorMessage));
+            IsSuccessful = false;
         }
     }
 
     public class GetResponse<T> : Response
     {
         public T RequestedData { get; set; }
+
+        public GetResponse(string ErrorMessage) : base(ErrorMessage) { }
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Models/Utils/Response.cs]: Why would you need generic response with error?

File: Cinema system/Cinema system/Models/Utils/Response.cs [hunk 16]
```diff
@@ -10,17 +12,50 @@ public class Response
...
     public class CreateResponse : Response
     {
         public int Id { get; set; }
+
+        public CreateResponse(int id)
+        {
+            IsSuccessful = true;
+            Id = id;
+        }
+        public CreateResponse(string ErrorMessage) : base(ErrorMessage) { }
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Models/Utils/Response.cs]: Why would you need CreateResponse with error. Shouldn't base class be enough?

File: Cinema system/Cinema system/Controllers/CinemasController.cs [hunk 17]
```diff
@@ -14,254 +12,171 @@ public class CinemasController : ControllerBase
...
         public IActionResult GetCinema(int cinemaId)
         {
-            var response = new GetResponse<CinemaView>();
-
             if (!_сinemaService.CheckCinemaExists(cinemaId))
             {
-                response.IsSuccessful = false;
-                response.ErrorDescriptions.Add("This cinema does not exists.");
-                return NotFound(response);
+                return NotFound(new GetResponse<CinemaView>(ErrorOfCinemaNonexistence));
```
Comment (andrei-kruhliakou-itechart) [inline Cinema system/Cinema system/Controllers/CinemasController.cs]: Why don't return base class?
Applies to all error responses in this controller.

**Review by andrei-kruhliakou-itechart — CHANGES_REQUESTED** [review]

File: Cinema system/Cinema system/Controllers/CinemasController.cs (hunk 12 above)
Comment (LozovskiM) [reply Cinema system/Cinema system/Controllers/CinemasController.cs]: It gives an error when I write :
new GetResponse(cinema)

**Review by andrei-kruhliakou-itechart — APPROVED** [review]

**Pull request merged**